src/SourceAdapters/StreamAdapter.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Plank\Mediable\SourceAdapters;
+
+use Plank\Mediable\Stream;
+use Plank\Mediable\Exceptions\MediaUpload\ConfigurationException;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Stream Adapter.
+ *
+ * Adapts a stream object or resource.
+ */
+abstract class StreamAdapter implements SourceAdapterInterface
+{
+    /**
+     * The source object.
+     * @var \Psr\Http\Message\StreamInterface
+     */
+    protected $source;
+
+    /**
+     * The resource.
+     * @var resource|null
+     */
+    protected $resource;
+
+    /**
+     * Constructor.
+     * @param \Psr\Http\Message\StreamInterface|resource $source
+     */
+    public function __construct($source)
+    {
+        if (is_resource($source) && get_resource_type($source) === 'stream') {
+            $this->resource = $source;
+            $this->source = new Stream($source);
+        } elseif ($source instanceof StreamInterface) {
+            $this->source = $source;
+        } else {
+            throw ConfigurationException::unrecognizedSource($source);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSource()
+    {
+        if ($this->resource) {
+            return $this->resource;
+        }
+
+        return $this->source;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function path()
+    {
+        return $this->source->getMetadata('uri');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function filename()
+    {
+        return pathinfo($this->path(), PATHINFO_FILENAME);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function extension()
+    {
+        return pathinfo($this->path(), PATHINFO_EXTENSION);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    abstract public function mimeType();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function contents()
```

Comment:
A similar check is done here because a `SourceInterface` return its contents as string, and there is no way to get the underlying resource.